Updating Identity Engine > Embedded Authentication content for Node.js SDK Recover Password use case (#2356)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/oie-embedded-sdk-use-cases/oie-embedded-sdk-use-case-pwd-recovery-mfa/nodejs/integrationsteps.md

## Integration steps

### Step 1: Add forgot password link to sign-in page

The first step is to create a forgot password link on the sign-in page, as in the following example. This link points to the "Reset your password" page.

<div class="common-image-format">

> ![Forgot password screenshot where user can click link to reset password.](/img/oie-embedded-sdk/oie-embedded-sdk-use-case-pwd-recovery-screenshot-forgot.png)

Author: Nice. Be good to review alt-text format as a team. Or at least I need to. Thanks.

</div>

Create a "Reset your password" page that initiates the reset password flow. This page accepts the user's email address and has a **Submit** button that starts the reset flow.

> After the user clicks **Submit** to start the reset flow, the next step is to call `idx.recoverPassword` and pass in the email address captured from the form. This starts the password recovery flow and returns a status of `Idx.Status:PENDING` and a `nextStep` field with inputs for an authenticator type.

Author: See comment later on

The call is shown in `recover-password.js` file of the SDK sample application.

```JavaScript
router.post('/recover-password', async (req, res, next) => {
  const { username } = req.body;
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx.recoverPassword({ username });
  handleTransaction({ req, res, next, authClient, transaction });
});
```

> The `handleTransaction` response, in the `handleTransaction.js` page, directs the user to a select authenticators page (`/select-authenticator` in this example) that the user needs to use to select an authentication type. In this case, the email factor is configured.

Author: I recast this bit to point to the correct page; the method is in another page and is referenced previously.

After the user selects the email authenticator, the next step is to call `idx.recoverPassword` with the authenticator type, as shown in the `authenticator.js` file of the SDK sample application. The method returns a status of `Idx.Status:PENDING`. This status indicates that the Okta platform has emailed the verification code to the user's email address and it's now awaiting verification. The response also includes a `nextStep` field that requires an email verification code input parameter.

```JavaScript
router.post('/select-authenticator', async (req, res, next) => {
  const { idxMethod } = req.getFlowStates();
  const { authenticator } = req.body;
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx[idxMethod]({ authenticator });
  handleTransaction({ req, res, next, authClient, transaction });
});
```

### Step 4: Create code verification page

> After the call to `idx.recoverPassword` has responded with a status of `Idx.Status:PENDING`, the user is redirected to a code verification page. The page must display a field to enter a code and a button or link to send the code to Okta for the email verification.

Reviewer: ```suggestion
After the call to `idx.recoverPassword()` has responded with a status of `Idx.Status:PENDING`, redirect the user to a code verification page. This page must display a field to enter a code and a button or link to send the code to Okta for the email verification.
```

Reviewer: `methodname()` format is normally used for calling out methods with parameters. But if this is not standard convention for node.js, please ignore this suggestion.

Author: I'm confirming. I've been simply stating to call it, and that's how I've done it all through my use-cases. I'll leave for now, but will clean this up on second pass.

> After the user checks their email for the code and enters the value into the field, they click a button to initiate the verification. Call `idx.recoverPassword` with the verification code passed in, as shown in the `authenticator.js` page of the SDK sample application.

Reviewer: ```suggestion
After the user checks their email for the code and enters the value into the field, they click a button to initiate the verification. Call the `idx.recoverPassword()` method with the verification code passed in, as shown in the `authenticator.js` page of the SDK sample application.
```

Author: See next comment

```JavaScript
router.post('/challenge-authenticator/email', async (req, res, next) => {
  const { idxMethod } = req.getFlowStates();
  const { verificationCode } = req.body;
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx[idxMethod]({ verificationCode });
  handleTransaction({ req, res, next, authClient, transaction });
});
```

> If the `idx.recoverPassword` call is successful, it returns a status of `Idx.Status:PENDING` and the `nextStep` field indicates an input of a password value parameter. This status indicates that the user can now change their password. At this point, the user is redirected to the change password page.

Reviewer: ```suggestion
If the `idx.recoverPassword()` call is successful, it returns a status of `Idx.Status:PENDING` and the `nextStep` field indicates an input of a password value parameter. This status indicates that the user can now change their password. At this point, the user is redirected to a change password page.
```

### Step 5: Create change password page

Create a change password page that allows the user to enter the new password and initiate the update.

> The final step is to make a call to `idx.recoverPassword`, passing in the value of the password, input by the user, as shown in the `recover-password.js` file of the SDK sample application. This call resets the user's password.

Reviewer: ```suggestion
The final step is to make a call to `idx.recoverPassword()`, passing in the value of the user input password, as shown in the `recover-password.js` file of the SDK sample application. This call resets the user's password.
```

```JavaScript
router.post('/reset-password', async (req, res, next) => {
  const { password, confirmPassword } = req.body;
  if (password !== confirmPassword) {
    next(new Error('Password not match'));
    return;
  }

  const authClient = getAuthClient(req);
  const transaction = await authClient.idx.recoverPassword({ password });
  handleTransaction({ req, res, next, authClient, transaction });
});
```

> If the response to the `idx.recoverPassword` call returns `Idx.Status:SUCCESS` and tokens, the change password flow completed successfully and you can redirect the user to the default home page.

Reviewer: ```suggestion
If the response to the `idx.recoverPassword()` call returns `Idx.Status:SUCCESS` and tokens, the change password flow completed successfully and you can redirect the user to the default home page.
```

Author: As per other comment on this terminology.
